utils/dataset_manifest/core.py

```diff
@@ -6,154 +6,192 @@
+                        raise Exception('Invalid dts sequences')
+                    frame_pts, frame_dts = frame.pts, frame.dts
+
                     if frame.key_frame:
-                        self._key_frames[frame_number] = {
+                        key_frame_number += 1
+                        ratio = (index + 1) // key_frame_number
+                        if ratio >= self._upper_bound and not self._force:
+                            raise AssertionError('Too few keyframes')
+                        key_frame = {
+                            'index': index,
                             'pts': frame.pts,
-                            'md5': md5_hash(frame),
+                            'md5': md5_hash(frame)
                         }
-                    frame_number += 1
-            self.frames = frame_number
-
-    @property
-    def key_frames(self):
-        return self._key_frames
-
-    def __len__(self):
-        return len(self._key_frames)
-
-    #TODO: need to change it in future
-    def __iter__(self):
-        for idx, key_frame in self._key_frames.items():
-            yield (idx, key_frame['pts'], key_frame['md5'])
 
+                        with closing(av.open(self.source_path, mode='r')) as checked_container:
+                            checked_container.seek(offset=key_frame['pts'], stream=video_stream)
+                            isValid = self.validate_key_frame(checked_container, video_stream, key_frame)
+                            if isValid:
+                                yield (index, key_frame['pts'], key_frame['md5'])
+                    index += 1
 
 class DatasetImagesReader:
-    def __init__(self, sources, meta=None, is_sorted=True, use_image_hash=False, *args, **kwargs):
+    def __init__(self,
+                sources,
+                meta=None,
+                is_sorted=True,
+                use_image_hash=False,
+                start = 0,
+                step = 1,
+                stop = None,
+                *args,
+                **kwargs):
         self._sources = sources if is_sorted else sorted(sources)
         self._meta = meta
-        self._content = []
         self._data_dir = kwargs.get('data_dir', None)
         self._use_image_hash = use_image_hash
+        self._start = start
+        self._stop = stop if stop else len(sources)
+        self._step = step
+
+    @property
+    def start(self):
+        return self._start
+
+    @start.setter
+    def start(self, value):
+        assert isinstance(value, int)
+        self._start = value
+
+    @property
+    def stop(self):
+        return self._stop
+
+    @stop.setter
+    def stop(self, value):
+        assert isinstance(value, int)
+        self._stop = value
+
+    @property
+    def step(self):
+        return self._step
+
+    @step.setter
+    def step(self, value):
+        assert isinstance(value, int)
+        self._step = value
```

Comment:
Probably better just `self._step = int(value)`. Thus if value can be converted to int, it will be converted. Otherwise you will get an exception (e.g. `int(1)`, `int('1')`, `int(1.5)`)